### PR TITLE
Use response synthesizer in context chat engines

### PR DESCRIPTION
**Pull request by logan-markewich**
The context chat engines do not handle token overflows well. It requires the user to
- limit the token limit on their memory
- limit the overall top-k on each retrieval

By using a response synthesizer, we can make this a little more painless.

This needed up being a decent refactor, but I think it simplified a lot too

### CODE REVIEW

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py [hunk 1]
```diff
@@ -38,6 +47,23 @@
...
+
+  Here are the relevant documents for the context:
+
+  {context_str}
+
+  Existing Answer:
+  {existing_answer}
+
+  Instruction: Refine the existing answer using the provided context to assist the user.
+  Answer "don't know" if not present in the document.
```
Comment (jerryjliu) [inline llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:64]: would this override an existing answer with don't know if the specifc new piece of context is not relevant?

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py [hunk 2]
```diff
@@ -150,41 +183,104 @@ async def _acondense_question(
...
+            text_qa_template=ChatPromptTemplate.from_messages(qa_messages),
+            refine_template=ChatPromptTemplate.from_messages(refine_messages),
+            streaming=streaming,
+        )
+
+    def _get_current_chat_history(
+        self, input_msg: str, nodes: List[NodeWithScore]
+    ) -> List[ChatMessage]:
+        # get the current memory buffer safely
+        if hasattr(self._memory, "tokenizer_fn"):
```
Comment (jerryjliu) [inline llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:260]: what is this for again

Comment (jerryjliu) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:260]: also i assume the memory doesn't include the system prompt?

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py [hunk 3]
```diff
@@ -150,41 +183,104 @@ async def _acondense_question(
...
         for postprocessor in self._node_postprocessors:
             nodes = postprocessor.postprocess_nodes(
                 nodes, query_bundle=QueryBundle(message)
             )
 
-        context_str = "\n\n".join(
-            [n.node.get_content(metadata_mode=MetadataMode.LLM).strip() for n in nodes]
+        return nodes
+
+    def _get_prefix_qa_messages_with_context(
```
Comment (jerryjliu) [inline llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:212]: no way to abstract this implementation + the contextchatengine impl. into one utils function? seems like mostly the same code

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py (hunk 1 above)
Comment (logan-markewich) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:64]: Good point, whoops

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py (hunk 2 above)
Comment (logan-markewich) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:260]: It's so that memory is retrieved taking the size of the retrieved context into account -- debatable if this is still needed or not

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py (hunk 3 above)
Comment (logan-markewich) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:212]: Hmm maybe, I can give it a shot

Comment (logan-markewich) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py]: addressed!

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py (hunk 1 above)
Comment (logan-markewich) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py]: Fixed!

File: llama-index-core/llama_index/core/chat_engine/condense_plus_context.py (hunk 2 above)
Comment (jerryjliu) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:242]: actually i'm not sure i understand, can you ELI5 why we limit memory based on retrieved context size

Comment (logan-markewich) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py:242]: So
- the user sets some token limit on the memory
- they would expect this to limit the input size of the memory buffer
- BUT the retrieved context is also added to the current memory buffer
- so, it will end up exceeding the token limit that you set
- a contributor added this functionality to make it possible to account for the retrieved context
- I carried this over from the previous version of the chat engine

Comment (logan-markewich) [reply llama-index-core/llama_index/core/chat_engine/condense_plus_context.py]: eh, just pulled out this concept, it was mostly a workaround to not using synthesizers before

**Review by enrico-stauss — COMMENTED** [review]
Resolved by d3992097a9c3e05591dd4caf96f2f8955a969d43

**Comment by shadyelgewily-slimstock** [comment]
Hi @logan-markewich, we just noticed this (seemingly) behavioral change by upgrading llama-index to the latest version. In previous versions of this class, there was no "refine" prompt. Is the additional refine prompt essential to preventing token overflows well? The implication for us is that we were unintentionally using the new variant of the class including the "refine" prompt, meaning extra tokens and suboptimal configuration (default refine prompt) in our production system.

As downstream users of the API, it's hard to be aware of behavioral changes like this, especially because llama-index has stopped providing detailed release notes since version v0.11.0). Are the release notes available somewhere else?

**Comment by logan-markewich** [comment]
@shadyelgewily-slimstock our changelog is up to date. Updating the github release notes is manual process and sometimes gets missed 

Before, there was no refine, and the chat engine would error out with a token/context window error. Now it doesn't

**Comment by shadyelgewily-slimstock** [comment]
Hi @logan-markewich, thanks for your response and I can indeed see the changelog is up to date. I will check there in the future. We also pinned the versions of all llama-index (sub) packages.

I would like to give the feedback that to us, "CondensePlusContext" has a specific meaning: it condenses a prompt and adds context from a knowledge base (potentially with self-managed chat history). I have looked through the code, and "CondensePlusContext" is now "CondensePlusContext" and "compactAndRefine" intermixed: condense a prompt, add context, and then do a (potentially iterative) refine and compact loop. To me, nothing in the name CondensePlusContext indicates that this will happen. So is it really still "CondensePlusContext" now? It seems almost like a compactAndRefine with additional condense prompt.

If we wanted to use compactAndRefine strategy, we would have used that strategy. It's your library and I know it is well before version 1 and we are simply users, so in the end it's your choice. But please do note that from a users' (and also from a potential contributor standpoint) perspective changing behavior to strategies that have particular meaning is very confusing. In this case, the "refine" part could have been optional while achieving the same goals.

**Comment by logan-markewich** [comment]
`CondensePlusContext` means `condense the chat history, use it for context retrieval, use the context for responding` -- the name is still representative of what it's doing imo?

Most people are surprised when token limits are encountered, because query engines automatically handle token limits, so why shouldn't a chat engine? That was the motivation for the change in any case. 

The refine step would only be hit in cases where an error would be raised previously, and with the current context window sizes on most LLMs, would be rarely hit

**Comment by shadyelgewily-slimstock** [comment]
Based on your response I checked what actually happened as the complete prompt (which does not exceed the token limits) enters the CompactAndRefine instance, and I indeed see that only the _text_qa_template is used when everything fits into a single token limit buffer. The refine prompt is only used when the buffer is exceeded.

Thanks for your answer and we resolved the issue on our end.
